orttraining/orttraining/core/graph/gradient_builder.cc

```diff
@@ -1613,5 +1614,74 @@ IMPLEMENT_GRADIENT_BUILDER(GetMinMaxGradient) {
   return result;
 }
 
+IMPLEMENT_GRADIENT_BUILDER(GetPythonOpGradient) {
+  std::vector<NodeDef> result;
+  auto src_attrs = SrcNodeAttributes();
+  std::vector<AttributeProto> attrs;
+  ORT_ENFORCE(utils::HasString(src_attrs.at("name")));
+  attrs.push_back(MakeAttribute("name", src_attrs.at("name").s()));
+
+  // input_tensor_types[i] store the type of autograd.Function.apply's ith output.
+  // Note that PythonOpGrad's 0-th input is the Python context generated by PythonOp.
+  std::vector<int64_t> input_tensor_types;
+  for (const auto input_type : src_attrs["output_tensor_types"].ints()) {
+    input_tensor_types.push_back(input_type);
+  }
+  attrs.push_back(MakeAttribute("input_tensor_types", input_tensor_types));
+
+  std::vector<int64_t> input_tensor_requires_grads;
+  // Context doesn't need gradient.
+  input_tensor_requires_grads.push_back(0);
+  // Set up gradients from outputs of autograd.Function.backward(...).
+  // The relation between forward and backward can be described by
+  // x -> forward -> y
+  // dy -> backward -> dx
+  // Here, if y (output of PythonOp) requires gradient, dy has requires_gradient=True.
+  for (const auto requires_grad : src_attrs["output_tensor_requires_grads"].ints()) {
+    input_tensor_requires_grads.push_back(requires_grad);
+  }
+  attrs.push_back(MakeAttribute("input_tensor_requires_grads", input_tensor_requires_grads));
+
+  // output_tensor_types[i] stores the type of autograd.Function.apply's ith input.
+  // We assume tensor and its gradient have the same type.
```

Comment:
We are not able to assert it because when exporting TorchIR to ONNX, the backward graph hasn't been generated yet.
On the other hand, we haven't seen any counter examples. I checked with Aishwarya and confirmed that, even for auto-cast, parameter and its gradietn have the same type.